### PR TITLE
Add basic serverside HUD elements

### PR DESCRIPTION
**Pull request by F1F7Y**
Adds 3 basic types of messages a server can display on a certain players HUD using string commands.
This means the client isn't required to be up-to-date to play, but to see the messages they need to update.

### Docs
All functions use a generic struct, not all use every field
```cpp
global struct genericServerHUDMessageStruct
{
	string title        = ""
	string description  = ""
	string image        = ""
	float duration      = 0
	vector color	= <0,0,0>
	int priority	= 0
	int style		= 0
}
```

#### `void function NSSendPopUpMessageToPlayer( entity player, genericServerHUDMessageStruct message )`
Only uses `genericServerHUDMessageStruct.description`

popup

#### `void function NSSendLargeMessageToPlayer( entity player, genericServerHUDMessageStruct message )`
Uses all fields of `genericServerHUDMessageStruct `

large

#### `void function NSSendInfoMessageToPlayer( entity player, genericServerHUDMessageStruct message )`
Only uses `genericServerHUDMessageStruct.description`

info

#### `void function NSSendAnnouncementMessageToPlayer( entity player, genericServerHUDMessageStruct message )`
Only uses `genericServerHUDMessageStruct.description` && `genericServerHUDMessageStruct.title`

announcement

### Status messages
There can only be 4 at a time, server can create, edit and delete them per player using an `id` it gets when it creates a message
Only uses `genericServerHUDMessageStruct.description` && `genericServerHUDMessageStruct.title`

#### `string function NSCreateStatusMessageOnPlayer( entity player, genericServerHUDMessageStruct message )`
Returns `id`
#### `void function NSDeleteStatusMessageOnPlayer( entity player, string id  )`

2status

### CODE REVIEW

**Review by BigSpice — APPROVED** [review]
Wanted to do custom kill streak ui.
well...i guess this is here now.
thank you @F1F7Y.

**Comment by x3Karma** [comment]
Could you add one more for sending announcement info such as the bankrupt announcement or infection announcement?

**Comment by F1F7Y** [comment]
ye :)

**Comment by F1F7Y** [comment]
Edited initial message with new stuff, ready for review

**Review by ASpoonPlaysGames — APPROVED** [review]

**Comment by x3Karma** [comment]
These are all very cool but I bet not many people will use the LargeMessage one simply because of the limited selection of background image.

**Comment by F1F7Y** [comment]
The large one occupies the same corner as the status messages so I hope it wont be overused. The main purpose for it is mainly like a welcome message. I tried adding image support, but couldn't get it to work.

**Comment by F1F7Y** [comment]
Large messages now support setting custom images

**Review by BobTheBob9 — CHANGES_REQUESTED** [review]

File: Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+{
+	ServerToClientStringCommand( player, "ServerHUDMessageSetDescription " + message.description )
+
+	ServerToClientStringCommand( player, "ServerHUDMessageRequest " + eMessageType.INFO + " " + message.duration )
+}
+#endif // SERVER
+
+
+#if CLIENT
+void function ServerCallback_SetServerTitle ( array< string > args  )
```
[the remark this answers is not in the document]
Comment (F1F7Y) [reply Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut]: image

[the remark this answers is not in the document]
Comment (GeckoEidechse) [reply Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut]: Resolved with 19aa9d00606bcd7da81f38685240a8a9f3ee2819

File: Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+
+		announcementMessageList.remove(0)
+	}
+}
+
+void function ShowAnnouncementMessage()
+{
+	AnnouncementData announcement = Announcement_Create( ServerMessage.title )
+	Announcement_SetSubText( announcement, ServerMessage.description )
+	Announcement_SetTitleColor( announcement, <1,1,0> )
```
Comment (EladNLG) [inline Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut:271]: Let us customize color please.

File: Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+	Announcement_SetPurge( announcement, true )
+	Announcement_SetPriority( announcement, 300 )
+	Announcement_SetSoundAlias( announcement, SFX_HUD_ANNOUNCE_QUICK )
+	Announcement_SetStyle( announcement, ANNOUNCEMENT_STYLE_QUICK )
```
Comment (EladNLG) [inline Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut:275]: Just, in general - let us customize a lot of this.

**Review by EladNLG — CHANGES_REQUESTED** [review]
Let modders customize the announcement messages, would be very nice.

**Comment by ASpoonPlaysGames** [comment]
Is there anything blocking this PR now? only change that's been requested seems outdated but is still appearing
image

File: Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+{
+	infoMessageList.append( clone ServerMessage )
+}
+
+#endif // CLIENT
+
+asset function NSAssetFromString( string str )
+{
+	return expect asset(compilestring(format("return $\"%s\"", str))())
+}
```
Comment (RoyalBlue1) [inline Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut:370]: since now StringToAsset is merged into Launcher this is not required anymore

**Review by RoyalBlue1 — CHANGES_REQUESTED** [review]

File: Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut (hunk 4 above)
Comment (uniboi) [reply Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut:370]: @F1F7Y I'm on vacation, so could you do this? :P

Comment (F1F7Y) [reply Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut:370]: Currently coming back from vacation, will push this when I have time

**Comment by catornot** [comment]
Make this work in SP, please.

File: Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut (hunk 4 above)
Comment (F1F7Y) [reply Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut]: done

**Comment by x3Karma** [comment]
Go even further beyond and allow servers to notify clients to add/modify icons RUI :trollface:

**Review by RoyalBlue1 — APPROVED** [review]

**Review by ASpoonPlaysGames — APPROVED** [review]
image
Seems good to merge, only blocking review seems outdated, so could probably be marked as resolved?

**Comment by uniboi** [comment]
@GeckoEidechse what's missing to merge this?

**Comment by GeckoEidechse** [comment]
> @GeckoEidechse what's missing to merge this?

Testing with POCs. Haven't gotten around to doing that yet.

**Comment by uniboi** [comment]
> > @GeckoEidechse what's missing to merge this?
> 
> Testing with POCs. Haven't gotten around to doing that yet.

I'm not a person of color but if you tell me what to do maybe I can help :P

**Comment by GeckoEidechse** [comment]
> > > @GeckoEidechse what's missing to merge this?
> > 
> > 
> > Testing with POCs. Haven't gotten around to doing that yet.
> 
> I'm not a person of color but if you tell me what to do maybe I can help :P

POC in this case means [proof of concept](https://en.wikipedia.org/wiki/Proof_of_concept) :P

Basically just checking out this branch and testing either with Squirrel commands in console or using a small sample mod that the above HUD elements actually work as intended. ^^

So yeah, feel free to give a shot and see if the stuff in this branch works in-game. If you do LMK what you tried and if it did (not) work <3

**Review by uniboi — APPROVED** [review]
Tested all of the different ruis and they work as expected

**Review by GeckoEidechse — APPROVED** [review]
Confirmed working in testing using the following demo mod

[GeckoEidechse.RuiDemo.zip](https://github.com/R2Northstar/NorthstarMods/files/9128643/GeckoEidechse.RuiDemo.zip)

**Comment by GeckoEidechse** [comment]
This would be ready to merge but I'm waiting on Bob's response in regards to Custom stuff outside of `Northstar.Custom`.

**Comment by GeckoEidechse** [comment]
> I know it's kinda late, but I'm considering redoing some parts and adding the cool source voting ui
> Should I close this and open a new PR when ready, or should I just add onto this @GeckoEidechse ?

Hmm, I'd keep it open but maybe set to draft mode, make new branch for your source voting stuff and then PR that new one. That way depending on when we'll release `v1.10.0` (I want like 1-2 more bigger features next to serverside HUD) we could either use this current one or your new version ^^

**Comment by GeckoEidechse** [comment]
Also depending on how much the overhauled version changes, e.g. if it just adds more functionality it'd might just make more sense to merge this and then the overhaul separately.

If you're gonna change some of the structs for example though it makes more sense to just put everything into the overhaul and then merge that ^^

**Comment by GeckoEidechse** [comment]
Superseded by #435

**Comment by GeckoEidechse** [comment]
Closing in favour of #435

**Pull request closed without merging**